src/mpn_extras/mul_toom22.c

```diff
@@ -0,0 +1,255 @@
+
+  * the GNU Lesser General Public License as published by the Free
+    Software Foundation; either version 3 of the License, or (at your
+    option) any later version.
+
+or
+
+  * the GNU General Public License as published by the Free Software
+    Foundation; either version 2 of the License, or (at your option) any
+    later version.
+
+or both in parallel, as here.
+
+The GNU MP Library is distributed in the hope that it will be useful, but
+WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+for more details.
+
+You should have received copies of the GNU General Public License and the
+GNU Lesser General Public License along with the GNU MP Library.  If not,
+see https://www.gnu.org/licenses/.  */
+
+#include "flint.h"
+#include "mpn_extras.h"
+
+/* Unoptimized */
+#define MPN_INCR_U(ptr, size, n)   mpn_add_1((ptr), (ptr), (size), (n))
+
+/* Evaluate in: -1, 0, +inf
+
+  <-s--><--n-->
+   ____ ______
+  |_a1_|___a0_|
+   |b1_|___b0_|
+   <-t-><--n-->
+
+  v0  =  a0     * b0       #   A(0)*B(0)
+  vm1 = (a0- a1)*(b0- b1)  #  A(-1)*B(-1)
+  vinf=      a1 *     b1   # A(inf)*B(inf)
+*/
+
+#define TOOM22_MUL_N_REC(p, a, b, n, ws) \
+    do { \
+        if (FLINT_HAVE_MUL_N_FUNC(n)) \
+            flint_mpn_mul_n_func_tab[n](p, a, b); \
+        else \
+            flint_mpn_mul_toom22(p, a, n, b, n, ws); \
+    } while (0)
+
+/* Normally, this calls mul_basecase or toom22_mul.  But when when the fraction
+   MUL_TOOM33_THRESHOLD / MUL_TOOM22_THRESHOLD is large, an initially small
+   relative unbalance will become a larger and larger relative unbalance with
+   each recursion (the difference s-t will be invariant over recursive calls).
+   Therefore, we need to call toom32_mul.  FIXME: Suppress depending on
+   MUL_TOOM33_THRESHOLD / MUL_TOOM22_THRESHOLD and on MUL_TOOM22_THRESHOLD.  */
+#define TOOM22_MUL_REC(p, a, an, b, bn, ws) \
+    do { \
+        if (FLINT_HAVE_MUL_FUNC(an, bn)) \
+            flint_mpn_mul_func_tab[an][bn](p, a, b); \
+        else if (bn <= 12 || 4 * an >= 5 * bn) \
+            flint_mpn_mul(p, a, an, b, bn); \
+        else \
+            flint_mpn_mul_toom22(p, a, an, b, bn, ws); \
+    } while (0)
+
+/*
+    else if (4 * an < 5 * bn)                        \
+      mpn_toom22_mul (p, a, an, b, bn, ws);                \
+    else                                \
+      mpn_toom32_mul (p, a, an, b, bn, ws);                \
+*/
+
+/* Needed scratch space is 2*(an + k), where k is the recursion depth. */
+
+void
+flint_mpn_mul_toom22(mp_ptr pp,
+        mp_srcptr ap, mp_size_t an,
+        mp_srcptr bp, mp_size_t bn,
+        mp_ptr scratch)
+{
+    mp_size_t n, s, t;
+    int vm1_neg;
+    mp_limb_t cy, cy2;
+    mp_ptr asm1;
+    mp_ptr bsm1;
+    TMP_INIT;
+
+    TMP_START;
+    if (scratch == NULL)
+    {
+        scratch = TMP_ALLOC(sizeof(mp_limb_t) * (2 * ((an) + FLINT_BITS)));
+    }
```

Comment:
I do not see any reason to why we would allow `an` and `bn` to be of any size, but I may be wrong. We only want to utilize Karutsuba when both `an` and `bn` are under some certain size, right? And this upper bound is not huge, correct? In this case we could just allocate it on the stack, no?

Reply:
Or, if `2 * (an + FLINT_BITS)` is greater than some size (say 256), do `TMP_ALLOC`, but try to avoid it?